File: applications/luci-app-dockerman/luasrc/model/cbi/dockerman/images.lua
--[[
LuCI - Lua Configuration Interface
Copyright 2019 lisaac <https://github.com/lisaac/luci-app-dockerman>
]]--

require "luci.util"
local uci = luci.model.uci.cursor()
local docker = require "luci.model.docker"
local dk = docker.new()

local containers, images
local res = dk.images:list()
if res.code <300 then images = res.body else return end
res = dk.containers:list({query = {all=true}})
if res.code <300 then containers = res.body else return end

function get_images()
  local data = {}
  for i, v in ipairs(images) do
    local index = v.Created .. v.Id
    data[index]={}
    data[index]["_selected"] = 0
    data[index]["id"] = v.Id:sub(8)
    data[index]["_id"] = '<a href="javascript:new_tag(\''..v.Id:sub(8,20)..'\')" class="dockerman-link" title="'..translate("New tag")..'">' .. v.Id:sub(8,20) .. '</a>'
    if v.RepoTags and next(v.RepoTags)~=nil then
      for i, v1 in ipairs(v.RepoTags) do
        data[index]["_tags"] =(data[index]["_tags"] and ( data[index]["_tags"] .. "<br>" )or "") .. ((v1:match("<none>") or (#v.RepoTags == 1)) and v1 or ('<a href="javascript:un_tag(\''..v1..'\')" class="dockerman_link" title="'..translate("Remove tag")..'" >' .. v1 .. '</a>'))
        if not data[index]["tag"] then
          data[index]["tag"] = v1--:match("<none>") and nil or v1
        end
      end
    else
      data[index]["_tags"] = v.RepoDigests[1] and v.RepoDigests[1]:match("^(.-)@.+")
      data[index]["_tags"] = (data[index]["_tags"] and data[index]["_tags"] or  "<none>" ).. ":<none>"
    end
    data[index]["_tags"] = data[index]["_tags"]:gsub("<none>","&lt;none&gt;")
    -- data[index]["_tags"] = '<a href="javascript:handle_tag(\''..data[index]["_id"]..'\')">' .. data[index]["_tags"] .. '</a>'
    for ci,cv in ipairs(containers) do
      if v.Id == cv.ImageID then
        data[index]["_containers"] = (data[index]["_containers"] and (data[index]["_containers"] .. " | ") or "")..
        '<a href='..luci.dispatcher.build_url("admin/docker/container/"..cv.Id)..' class="dockerman_link" title="'..translate("Container detail")..'">'.. cv.Names[1]:sub(2).."</a>"
      end
    end
    data[index]["_size"] = string.format("%.2f", tostring(v.Size/1024/1024)).."MB"
    data[index]["_created"] = os.date("%Y/%m/%d %H:%M:%S",v.Created)
  end
  return data
end

local image_list = get_images()

-- m = Map("docker", translate("Docker"))
m = SimpleForm("docker", translate("Docker"))
m.submit=false
m.reset=false

local pull_value={_image_tag_name="", _registry="index.docker.io"}
local pull_section = m:section(SimpleSection, translate("Pull Image"))
pull_section.template="cbi/nullsection"
local tag_name = pull_section:option(Value, "_image_tag_name")
tag_name.template = "dockerman/cbi/inlinevalue"
tag_name.placeholder="lisaac/luci:latest"
local action_pull = pull_section:option(Button, "_pull")
action_pull.inputtitle= translate("Pull")
action_pull.template = "dockerman/cbi/inlinebutton"
action_pull.inputstyle = "add"
tag_name.write = function(self, section, value)
  local hastag = value:find(":")
  if not hastag then
    value = value .. ":latest"
  end
  pull_value["_image_tag_name"] = value
end
action_pull.write = function(self, section)
  local tag = pull_value["_image_tag_name"]
  local json_stringify = luci.jsonc and luci.jsonc.stringify
  if tag and tag ~= "" then
    docker:write_status("Images: " .. "pulling" .. " " .. tag .. "...\n")
    -- local x_auth = nixio.bin.b64encode(json_stringify({serveraddress= server})) , header={["X-Registry-Auth"] = x_auth}
    local res = dk.images:create({query = {fromImage=tag}}, docker.pull_image_show_status_cb)
    -- {"errorDetail": {"message": "failed to register layer: ApplyLayer exit status 1 stdout:  stderr: write \/docker: no space left on device" }, "error": "failed to register layer: ApplyLayer exit status 1 stdout:  stderr: write \/docker: no space left on device" }
    if res and res.code == 200 and (res.body[#res.body] and not res.body[#res.body].error and res.body[#res.body].status and (res.body[#res.body].status == "Status: Downloaded newer image for ".. tag)) then
      docker:clear_status()
    else
      docker:append_status("code:" .. res.code.." ".. (res.body[#res.body] and res.body[#res.body].error or (res.body.message or res.message)).. "\n")
    end
  else
    docker:append_status("code: 400 please input the name of image name!")
  end
  luci.http.redirect(luci.dispatcher.build_url("admin/docker/images"))
end

local import_section = m:section(SimpleSection, translate("Import Images"))
local im = import_section:option(DummyValue, "_image_import")
im.template = "dockerman/images_import"

local image_table = m:section(Table, image_list, translate("Images"))

local image_selecter = image_table:option(Flag, "_selected","")
image_selecter.disabled = 0
image_selecter.enabled = 1
image_selecter.default = 0

local image_id = image_table:option(DummyValue, "_id", translate("ID"))
image_id.rawhtml = true
image_table:option(DummyValue, "_tags", translate("RepoTags")).rawhtml = true
image_table:option(DummyValue, "_containers", translate("Containers")).rawhtml = true
image_table:option(DummyValue, "_size", translate("Size"))
image_table:option(DummyValue, "_created", translate("Created"))
image_selecter.write = function(self, section, value)
  image_list[section]._selected = value
end

local remove_action = function(force)
  local image_selected = {}
  -- 遍历table中sectionid
  local image_table_sids = image_table:cfgsections()
  for _, image_table_sid in ipairs(image_table_sids) do
    -- 得到选中项的名字
    if image_list[image_table_sid]._selected == 1 then
      image_selected[#image_selected+1] = (image_list[image_table_sid]["_tags"]:match("<br>") or image_list[image_table_sid]["_tags"]:match("&lt;none&gt;")) and image_list[image_table_sid].id or image_list[image_table_sid].tag
    end
  end
  if next(image_selected) ~= nil then
    local success = true
    docker:clear_status()
    for _,img in ipairs(image_selected) do
      docker:append_status("Images: " .. "remove" .. " " .. img .. "...")
      local query
      if force then query = {force = true} end
      local msg = dk.images:remove({id = img, query = query})
      if msg.code ~= 200 then
        docker:append_status("code:" .. msg.code.." ".. (msg.body.message and msg.body.message or msg.message).. "\n")
        success = false
      else
        docker:append_status("done\n")
      end
    end
    if success then docker:clear_status() end
    luci.http.redirect(luci.dispatcher.build_url("admin/docker/images"))
  end
end

local docker_status = m:section(SimpleSection)
docker_status.template = "dockerman/apply_widget"
docker_status.err = docker:read_status()
docker_status.err = docker_status.err and docker_status.err:gsub("\n","<br>"):gsub(" ","&nbsp;")
if docker_status.err then docker:clear_status() end

local action = m:section(Table,{{}})
action.notitle=true
action.rowcolors=false
action.template="cbi/nullsection"

local btnremove = action:option(Button, "remove")
btnremove.inputtitle= translate("Remove")
btnremove.template = "dockerman/cbi/inlinebutton"
btnremove.inputstyle = "remove"
btnremove.forcewrite = true
btnremove.write = function(self, section)
  remove_action()
end

local btnforceremove = action:option(Button, "forceremove")
btnforceremove.inputtitle= translate("Force Remove")
btnforceremove.template = "dockerman/cbi/inlinebutton"
btnforceremove.inputstyle = "remove"
btnforceremove.forcewrite = true
btnforceremove.write = function(self, section)
  remove_action(true)
end

local btnsave = action:option(Button, "save")
btnsave.inputtitle= translate("Save")
btnsave.template = "dockerman/cbi/inlinebutton"
btnsave.inputstyle = "edit"
btnsave.forcewrite = true
btnsave.write = function (self, section)
  local image_selected = {}
  local image_table_sids = image_table:cfgsections()
  for _, image_table_sid in ipairs(image_table_sids) do
    if image_list[image_table_sid]._selected == 1 then
      image_selected[#image_selected+1] = image_list[image_table_sid].id --image_id:cfgvalue(image_table_sid)
    end
  end
  if next(image_selected) ~= nil then
    local names
    for _,img in ipairs(image_selected) do
      names = names and (names .. "&names=".. img) or img
    end
    local first
    local cb = function(res, chunk)
      if res.code == 200 then
        if not first then
          first = true
          luci.http.header('Content-Disposition', 'inline; filename="images.tar"')
          luci.http.header('Content-Type', 'application\/x-tar')
        end
        luci.ltn12.pump.all(chunk, luci.http.write)
      else
        if not first then
          first = true
          luci.http.prepare_content("text/plain")
        end
        luci.ltn12.pump.all(chunk, luci.http.write)
      end
    end
    docker:write_status("Images: " .. "save" .. " " .. table.concat(image_selected, "\n") .. "...")
    local msg = dk.images:get({query = {names = names}}, cb)
    if msg.code ~= 200 then
      docker:append_status("code:" .. msg.code.." ".. (msg.body.message and msg.body.message or msg.message).. "\n")
      success = false
    else
      docker:clear_status()
    end
  end
end

local btnload = action:option(Button, "load")
btnload.inputtitle= translate("Load")
btnload.template = "dockerman/images_load"
btnload.inputstyle = "add"
return m

File: applications/luci-app-dockerman/htdocs/luci-static/resources/view/dockerman/images.js
'use strict';
'require form';
'require fs';
'require poll';
'require ui';
'require dockerman.common as dm2';

/*
Copyright 2026
Docker manager JS for Luci by Paul Donald <newtwen+github@gmail.com> 
Based on Docker Lua by lisaac <https://github.com/lisaac/luci-app-dockerman>
LICENSE: GPLv2.0
*/


return dm2.dv.extend({
	load() {
		return Promise.all([
			dm2.image_list(),
			dm2.container_list({query: {all: true}}),
		])
	},

	render([images, containers]) {
		if (images?.code !== 200) {
			return E('div', {}, [ images.body.message ]);
		}

		let image_list = this.getImagesTable(images.body);
		let container_list = containers.body;
		const view = this; // Capture the view context
		view.selectedImages = {};

		let s, o;
		const m = new form.JSONMap({image: image_list, pull: {}, push: {}, build: {}, import: {}, prune: {}},
			_('Docker - Images'),
			_('On this page all images are displayed that are available on the system and with which a container can be created.'));
		m.submit = false;
		m.reset = false;

		let pollPending = null;
		let imgSec = null;
		const calculateSizeTotal = () => {
			return Array.isArray(image_list) ? image_list.map(c => c?.Size).reduce((acc, e) => acc + e, 0) : 0;
		};

		const refresh = () => {
			if (pollPending) return pollPending;
			pollPending = view.load().then(([images2, containers2]) => {
				image_list = view.getImagesTable(images2.body);
				container_list = containers2.body;
				m.data = new m.data.constructor({ image: image_list, pull: {}, push: {}, build: {}, import: {}, prune: {} });

				const size_total = calculateSizeTotal();
				if (imgSec) {
					imgSec.footer = [
						'',
						`${_('Total')} ${image_list.length}`,
						'',
						`${'%1024mB'.format(size_total)}`,
					];
				}

				return m.render();
			}).catch((err) => { console.warn(err) }).finally(() => { pollPending = null });
			return pollPending;
		};

		// Pull image

		s = m.section(form.TableSection, 'pull',  dm2.Types['image'].sub['pull'].i18n,
			_('By entering a valid image name with the corresponding version, the docker image can be downloaded from the configured registry.'));
		s.anonymous = true;
		s.addremove = false;

		const splitImageTag = (value) => {
			const input = String(value || '').trim();
			if (!input || input.includes(' ')) return { name: '', tag: 'latest' };

			const lastSlash = input.lastIndexOf('/');
			const lastColon = input.lastIndexOf(':');
			if (lastColon > lastSlash) {
				return {
					name: input.slice(0, lastColon) || input,
					tag: input.slice(lastColon + 1) || 'latest'
				};
			}

			return { name: input, tag: 'latest' };
		};

		let tagOpt = s.option(form.Value, '_image_tag_name');
		tagOpt.placeholder = "[registry.io[:443]/]foobar/product:latest";

		o = s.option(form.Button, '_pull');
		o.inputtitle = `${dm2.Types['image'].sub['pull'].i18n} ${dm2.Types['image'].sub['pull'].e}`; // _('Pull') + ' ☁️⬇️'
		o.inputstyle = 'add';
		o.onclick = L.bind(function(ev, btn) {
			const raw = tagOpt.formvalue('pull') || '';
			const input = String(raw).trim();
			if (!input) {
				ui.addTimeLimitedNotification(dm2.Types['image'].sub['pull'].i18n, _('Please enter an image tag'), 4000, 'warning');
				return false;
			}

			const { name, tag: ver } = splitImageTag(input);

			return this.super('handleXHRTransfer', [{
				q_params: { query: { fromImage: name, tag: ver } },
				commandCPath: `/images/create`,
				commandDPath: `/images/create`,
				commandTitle: dm2.Types['image'].sub['pull'].i18n,
				successMessage: _('Image create completed'),
				onUpdate: (msg) => {
					try {
						if(msg.error)
							ui.addTimeLimitedNotification(dm2.ActionTypes['build'].i18n, msg.error, 7000, 'error');

						const output = JSON.stringify(msg, null, 2) + '\n';
						view.insertOutput(output);
					} catch {

					}
				},
				onSuccess: () => refresh(),
				noFileUpload: true,
			}]);

			// return view.executeDockerAction(
			// 	dm2.image_create,
			// 	{ query: { fromImage: name, tag: ver } },
			// 	dm2.Types['image'].sub['pull'].i18n,
			// 	{
			// 		showOutput: true,
			// 		successMessage: _('Image create completed')
			// 	}
			// );
		}, this);

		// Push image

		s = m.section(form.TableSection, 'push',  dm2.Types['image'].sub['push'].i18n,
			_('Push an image to a registry. Select an image tag from all available tags on the system.'));
		s.anonymous = true;
		s.addremove = false;

		// Build a list of all available tags across all images
		const allImageTags = [];
		for (const image of image_list) {
			const tags = Array.isArray(image.RepoTags) ? image.RepoTags : [];
			for (const tag of tags) {
				if (tag && tag !== '<none>:<none>') {
					allImageTags.push(tag);
				}
			}
		}

		let pushTagOpt = s.option(form.Value, '_image_tag_push');
		pushTagOpt.placeholder = _('Select image tag');
		if (allImageTags.length === 0) {
			pushTagOpt.value('', _('No image tags available'));
		} else {
			// Add all unique tags to the dropdown
			const uniqueTags = [...new Set(allImageTags)].sort();
			for (const tag of uniqueTags) {
				pushTagOpt.value(tag, tag);
			}
		}

		o = s.option(form.Button, '_push');
		o.inputtitle = `${dm2.Types['image'].sub['push'].i18n} ${dm2.Types['image'].sub['push'].e}`; // _('Push') + ' ☁️⬆️'
		o.inputstyle = 'add';
		o.onclick = L.bind(function(ev, btn) {
			const selected = pushTagOpt.formvalue('push') || '';
			if (!selected) {
				ui.addTimeLimitedNotification(dm2.Types['image'].sub['push'].i18n, _('Please select an image tag to push'), 4000, 'warning');
				return false;
			}

			const { name, tag: ver } = splitImageTag(selected);

			return this.super('handleXHRTransfer', [{
				// Pass name in q_params to trigger building X-Registry-Auth header
				q_params: { name: name, query: { tag: ver } },
				commandCPath: `/images/push/${name}`,
				commandDPath: `/images/${name}/push`,
				commandTitle: dm2.Types['image'].sub['push'].i18n,
				successMessage: _('Image push completed'),
				onSuccess: () => refresh(),
				noFileUpload: true,
			}]);

			// return view.executeDockerAction(
			// 	dm2.image_push,
			// 	{ name: name, query: { tag: ver} },
			// 	dm2.Types['image'].sub['push'].i18n,
			// 	{
			// 		showOutput: true,
			// 		successMessage: _('Image push completed')
			// 	}
			// );
		}, this);


		s = m.section(form.TableSection, 'build',  dm2.ActionTypes['build'].i18n,
			_('Build an image.') + ' ' + _('git repositories require git installed on the docker host.'));
		s.anonymous = true;
		s.addremove = false;

		let buildOpt = s.option(form.Value, '_image_build_uri');
		buildOpt.placeholder = "https://host/foo/bar.git | https://host/foobar.tar";

		let buildTagOpt = s.option(form.Value, '_image_build_tag');
		buildTagOpt.placeholder = 'repository:tag';

		o = s.option(form.Button, '_build');
		o.inputtitle = `${dm2.ActionTypes['build'].i18n} ${dm2.ActionTypes['build'].e}`; // _('Build') + ' 🏗️'
		o.inputstyle = 'add';
		o.onclick = L.bind(function(ev, btn) {
			const uri = buildOpt.formvalue('build') || '';
			const t = buildTagOpt.formvalue('build') || '';

			const q_params = { q: encodeURIComponent('false'), t: t };
			if (uri) q_params.remote = encodeURIComponent(uri);

			return this.super('handleXHRTransfer', [{
				q_params: { query: q_params },
				commandCPath: '/images/build',
				commandDPath: '/build',
				commandTitle: dm2.ActionTypes['build'].i18n,
				successMessage: _('Image loaded successfully'),
				onUpdate: (msg) => {
					try {
						if(msg.error)
							ui.addTimeLimitedNotification(dm2.ActionTypes['build'].i18n, msg.error, 7000, 'error');

						const output = JSON.stringify(msg, null, 2) + '\n';
						view.insertOutput(output);
					} catch {

					}
				},
				onSuccess: () => refresh(),
				noFileUpload: !!uri,
			}]);
		}, this);

		o = s.option(form.Button, '_delete_cache', null);
		o.inputtitle = `${dm2.ActionTypes['clean'].i18n} ${dm2.ActionTypes['clean'].e}`;
		o.inputstyle = 'negative';
		o.onclick = L.bind(function(ev, btn) {
			return this.super('handleXHRTransfer', [{
				q_params: { query: { all: 'true' } },
				commandCPath: '/images/build/prune',
				commandDPath: '/build/prune',
				commandTitle: dm2.Types['builder'].sub['prune'].i18n,
				successMessage: _('Cleaned build cache'),
				onUpdate: (msg) => {
					try {
						if(msg.error)
							ui.addTimeLimitedNotification(dm2.ActionTypes['clean'].i18n, msg.error, 7000, 'error');

						const output = JSON.stringify(msg, null, 2) + '\n';
						view.insertOutput(output);
					} catch {

					}
				},
				noFileUpload: true,
			}]);
		}, this);

		// Import image

		s = m.section(form.TableSection, 'import', dm2.Types['image'].sub['import'].i18n,
			_('Download a valid remote image tar.'));
		s.addremove = false;
		s.anonymous = true;

		let imgsrc = s.option(form.Value, '_image_source');
		imgsrc.placeholder = 'https://host/image.tar';

		let tagimpOpt = s.option(form.Value, '_import_image_tag_name');
		tagimpOpt.placeholder = 'repository:tag';

		let importBtn = s.option(form.Button, '_import');
		importBtn.inputtitle = `${dm2.Types['image'].sub['import'].i18n} ${dm2.Types['image'].sub['import'].e}` //_('Import') + ' ➡️';
		importBtn.inputstyle = 'add';
		importBtn.onclick = L.bind(function(ev, btn) {
			const rawtag = tagimpOpt.formvalue('import') || '';
			const input = String(rawtag).trim();
			if (!input) {
				ui.addTimeLimitedNotification(dm2.Types['image'].sub['import'].i18n, _('Please enter an image repo tag'), 4000, 'warning');
				return false;
			}
			const rawremote = imgsrc.formvalue('import') || '';
			let remote = String(rawremote).trim();
			if (!remote) {
				ui.addTimeLimitedNotification(dm2.Types['image'].sub['import'].i18n, _('Please enter an image source'), 4000, 'warning');
				return false;
			}

			const { name, tag: ver } = splitImageTag(input);

			return this.super('handleXHRTransfer', [{
				q_params: { query: { fromSrc: remote, repo: ver } },
				commandCPath: '/images/create',
				commandDPath: '/images/create',
				commandTitle: dm2.Types['image'].sub['create'].i18n,
				onUpdate: (msg) => {
					try {
						if(msg.error)
							ui.addTimeLimitedNotification(dm2.Types['image'].sub['create'].i18n, msg.error, 7000, 'error');

						const output = JSON.stringify(msg, null, 2) + '\n';
						view.insertOutput(output);
					} catch {

					}
				},
				onSuccess: () => refresh(),
				noFileUpload: true,
			}]);

			// return view.executeDockerAction(
			// 	dm2.image_create,
			// 	{ query: { fromSrc: remote, repo: ver } },
			// 	dm2.Types['image'].sub['import'].i18n,
			// 	{
			// 		showOutput: true,
			// 		successMessage: _('Image create started/completed')
			// 	}
			// );
		}, this);


		s = m.section(form.TableSection, 'prune', _('Images overview'), );
		s.addremove = false;
		s.anonymous = true;

		const prune = s.option(form.Button, '_prune', null);
		prune.inputtitle = `${dm2.ActionTypes['prune'].i18n} ${dm2.ActionTypes['prune'].e}`;
		prune.inputstyle = 'negative';
		prune.onclick = L.bind(function(ev, btn) {

			return this.super('handleXHRTransfer', [{
				q_params: {  },
				commandCPath: '/images/prune',
				commandDPath: '/images/prune',
				commandTitle: dm2.ActionTypes['prune'].i18n,
				onUpdate: (msg) => {
					try {
						if(msg.error)
							ui.addTimeLimitedNotification(dm2.ActionTypes['prune'].i18n, msg.error, 7000, 'error');

						const output = JSON.stringify(msg, null, 2) + '\n';
						view.insertOutput(output);
					} catch {

					}
				},
				onSuccess: () => refresh(),
				noFileUpload: true,
			}]);

			// return view.executeDockerAction(
			// 	dm2.image_prune,
			// 	{ query: { filters: '' } },
			// 	dm2.ActionTypes['prune'].i18n,
			// 	{
			// 		showOutput: true,
			// 		successMessage: _('started/completed'),
			//		onSuccess: () => refresh(),
			// 	}
			// );
		}, this);

		o = s.option(form.Button, '_export', null);
		o.inputtitle = `${dm2.ActionTypes['save'].i18n} ${dm2.ActionTypes['save'].e}`;
		o.inputstyle = 'cbi-button-positive';
		o.onclick = L.bind(function(ev, btn) {
			ev.preventDefault();

			const selected = Object.keys(view.selectedImages).filter(k => view.selectedImages[k]);
			if (!selected.length) {
				ui.addTimeLimitedNotification(_('Export'), _('No images selected'), 3000, 'warning');
				return;
			}

			// Get tags or IDs for selected images
			const names = selected.map(sid => {
				const image = s.map.data.data[sid];
				const tag = image?.RepoTags?.[0];
				return tag || image?.Id?.substr(12);
			});

			// http.uc does not yet handle parameter arrays, so /images/get needs access to the URL params
			window.location.href = `${view.dockerman_url}/images/get?${names.map(e => `names=${e}`).join('&')}`;

		}, this);

		const size_total = calculateSizeTotal();

		imgSec = m.section(form.TableSection, 'image');
		imgSec.anonymous = true;
		imgSec.nodescriptions = true;
		imgSec.addremove = true;
		imgSec.sortable = true;
		imgSec.filterrow = true;
		imgSec.addbtntitle = `${dm2.ActionTypes['upload'].i18n} ${dm2.ActionTypes['upload'].e}`;
		imgSec.footer = [
			'',
			`${_('Total')} ${image_list.length}`,
			'',
			`${'%1024mB'.format(size_total)}`,
		];

		imgSec.handleAdd = function(sid, ev) {
			return view.handleFileUpload();
		};

		imgSec.handleGet = function(image, ev) {
			const tag = image.RepoTags?.[0];
			const name = tag || image.Id.substr(12);

			// Direct HTTP download - avoid RPC
			window.location.href = `${view.dockerman_url}/images/get/${name}`;
			return true;
		};

		imgSec.handleRemove = function(sid, image, force=false, ev) {
			return view.executeDockerAction(
				dm2.image_remove,
				{ id: image.Id, query: { force: force } },
				dm2.ActionTypes['remove'].i18n,
				{
					showOutput: true,
					onSuccess: () => {
						delete this.map.data.data[sid];
						return this.super('handleRemove', [ev]);
					}
				}
			);
		};

		imgSec.handleInspect = function(image, ev) {
			return view.executeDockerAction(
				dm2.image_inspect,
				{ id: image.Id },
				dm2.ActionTypes['inspect'].i18n,
				{ showOutput: true, showSuccess: false }
			);
		};

		imgSec.handleHistory = function(image, ev) {
			return view.executeDockerAction(
				dm2.image_history,
				{ id: image.Id },
				dm2.ActionTypes['history'].i18n,
				{ showOutput: true, showSuccess: false }
			);
		};

		imgSec.renderRowActions = function (sid) {
			const image = this.map.data.data[sid];
			const btns = [
				E('button', {
					'class': 'cbi-button cbi-button-neutral',
					'title': dm2.ActionTypes['inspect'].i18n,
					'click': ui.createHandlerFn(this, this.handleInspect, image),
				}, [dm2.ActionTypes['inspect'].e]),
				E('button', {
					'class': 'cbi-button cbi-button-neutral',
					'title': dm2.ActionTypes['history'].i18n,
					'click': ui.createHandlerFn(this, this.handleHistory, image),
				}, [dm2.ActionTypes['history'].e]),
				E('button', {
					'class': 'cbi-button cbi-button-positive save',
					'title': dm2.ActionTypes['save'].i18n,
					'click': ui.createHandlerFn(this, this.handleGet, image),
				}, [dm2.ActionTypes['save'].e]),
				E('div', {
					'style': 'width: 20px',
					// Some safety margin for mis-clicks
				}, [' ']),
				E('button', {
					'class': 'cbi-button cbi-button-negative remove',
					'title': dm2.ActionTypes['remove'].i18n,
					'click': ui.createHandlerFn(this, this.handleRemove, sid, image, false),
					'disabled': image?._disable_delete,
				}, [dm2.ActionTypes['remove'].e]),
				E('button', {
					'class': 'cbi-button cbi-button-negative important remove',
					'title': dm2.ActionTypes['force_remove'].i18n,
					'click': ui.createHandlerFn(this, this.handleRemove, sid, image, true),
					'disabled': image?._disable_delete,
				}, [dm2.ActionTypes['force_remove'].e]),
			];
			return E('td', { 'class': 'td middle cbi-section-actions' }, E('div', btns));
		};

		o = imgSec.option(form.Flag, '_selected');
		o.onchange = function(ev, sid, value) {
			if (value == 1) {
				view.selectedImages[sid] = value;
			}
			else {
				delete view.selectedImages[sid];
			}
			return;
		}

		o = imgSec.option(form.DummyValue, 'RepoTags', dm2.Types['image'].sub['tag'].e);
		o.cfgvalue = function(sid) {
			const image = this.map.data.data[sid];
			const tags = Array.isArray(image?.RepoTags) ? image.RepoTags : [];

			if (tags.length === 0 || (tags.length === 1 && tags[0] === '<none>:<none>'))
				return '<none>';

			const tagLinks = tags.map(tag => {
				if (tag === '<none>:<none>')
					return E('span', {}, tag);

				/* last tag - don't link it - last tag removal == delete */
				if (tags.length === 1)
					return tag;

				return E('a', {
					'href': '#',
					'title': _('Click to remove this tag'),
					'click': ui.createHandlerFn(view, (tag, imageId, ev) => {

						ev.preventDefault();
						ui.showModal(_('Remove tag'), [
							E('p', {}, _('Do you want to remove the tag "%s"?').format(tag)),
							E('div', { 'class': 'right' }, [
								E('button', {
									'class': 'cbi-button',
									'click': ui.hideModal
								}, '↩'),
								' ',
								E('button', {
									'class': 'cbi-button cbi-button-negative',
									'click': ui.createHandlerFn(view, () => {
										ui.hideModal();

										return view.executeDockerAction(
											dm2.image_remove,
											{ id: tag, query: { noprune: 'true' } },
											dm2.Types['image'].sub['untag'].i18n,
											{
												showOutput: true,
												successMessage: _('Tag removed successfully'),
												successDuration: 4000,
												onSuccess: () => refresh(),
											}
										);
									})
								}, dm2.Types['image'].sub['untag'].e)
							])
						]);
					}, tag, image.Id)
				}, tag);
			});

			// Join with commas and spaces
			const content = [];
			for (const [i, tag] of tagLinks.entries()) {
				if (i > 0) content.push(', ');
				content.push(tag);
			}

			return E('span', {}, content);
		};

		o = imgSec.option(form.DummyValue, 'Containers', _('Containers'));
		o.cfgvalue = function(sid) {
			const imageId = this.map.data.data[sid].Id;
			// Collect all matching container name links for this image
			const anchors = container_list.reduce((acc, container) => {
				if (container?.ImageID !== imageId) return acc;
				for (const name of container?.Names || [])
					acc.push(E('a', { href: `container/${container.Id}` }, [ name.substring(1) ]));
				return acc;
			}, []);

			// Interleave separators
			if (!anchors.length) return E('div', {});
			const content = [];
			for (let i = 0; i < anchors.length; i++) {
				if (i) content.push(' | ');
				content.push(anchors[i]);
			}

			return E('div', {}, content);
		};

		o = imgSec.option(form.DummyValue, 'Size', _('Size'));
		o.cfgvalue = function(sid) {
			const s = this.map.data.data[sid].Size;
			return '%1024mB'.format(s);
		};
		imgSec.option(form.DummyValue, 'Created', _('Created'));
		o = imgSec.option(form.DummyValue, '_id', _('ID'));

		/* Remember: we load a JSONMap - so uci config is non-existent for these
		elements, so we must pull from this.map.data, otherwise o.load returns nothing */
		o.cfgvalue = function(sid) {
			const image = this.map.data.data[sid];
			const shortId = image?._id || '';
			const fullId = image?.Id || '';

			return E('a', {
				'href': '#',
				'style': 'font-family: monospace',
				'title': _('Click to add a new tag to this image'),
				'click': ui.createHandlerFn(view, function(imageId, ev) {
					ev.preventDefault();

					let repoInput, tagInput;
					ui.showModal(_('New tag'), [
						E('p', {}, _('Enter a new tag for image %s:').format(imageId.slice(7, 19))),
						E('div', { 'class': 'cbi-value' }, [
							E('label', { 'class': 'cbi-value-title' }, _('Repository')),
							E('div', { 'class': 'cbi-value-field' }, [
								repoInput = E('input', {
									'type': 'text',
									'class': 'cbi-input-text',
									'placeholder': '[registry.io[:443]/]myrepo/myimage'
								})
							])
						]),
						E('div', { 'class': 'cbi-value' }, [
							E('label', { 'class': 'cbi-value-title' }, _('Tag')),
							E('div', { 'class': 'cbi-value-field' }, [
								tagInput = E('input', {
									'type': 'text',
									'class': 'cbi-input-text',
									'placeholder': 'latest',
									'value': 'latest'
								})
							])
						]),
						E('div', { 'class': 'right' }, [
							E('button', {
								'class': 'cbi-button',
								'click': ui.hideModal
							}, ['↩']),
							' ',
							E('button', {
								'class': 'cbi-button cbi-button-positive',
								'click': ui.createHandlerFn(view, () => {
									const repo = repoInput.value.trim();
									const tag = tagInput.value.trim() || 'latest';

									if (!repo) {
										ui.addTimeLimitedNotification(null, [_('Repository cannot be empty')], 3000, 'warning');
										return;
									}

									ui.hideModal();

									return view.executeDockerAction(
										dm2.image_tag,
										{ id: imageId, query: { repo: repo, tag: tag } },
										dm2.Types['image'].sub['tag'].i18n,
										{
											showOutput: true,
											successMessage: _('Tag added successfully'),
											successDuration: 4000,
											onSuccess: () => refresh(),
										}
									);
								})
							}, [dm2.Types['image'].sub['tag'].e])
						])
					]);
				}, fullId)
			}, shortId);
		};

		this.insertOutputFrame(s, m);

		poll.add(L.bind(() => { refresh(); }, this), 10);

		return m.render();
	},

	handleFileUpload() {
		// const uploadUrl = `?quiet=${encodeURIComponent('false')}`;

		return this.super('handleXHRTransfer', [{
			q_params: { query: { quiet: 'false' } },
			commandCPath: `/images/load`,
			commandDPath: `/images/load`,
			commandTitle: _('Uploading…'),
			commandMessage: _('Uploading image…'),
			successMessage: _('Image loaded successfully'),
			defaultPath: '/tmp'
		}]);
	},

	handleSave: null,
	handleSaveApply: null,
	handleReset: null,

	getImagesTable(images) {
		const data = [];

		for (const image of images) {
			// Just push plain data objects without UCI metadata
			data.push({
				...image,
				_disable_delete: null,
				_id: (image.Id || '').substring(7, 20),
				Created: this.buildTimeString(image.Created) || '',
			});
		}

		return data;
	},

});
